### DISCUSSION TITLE
get list of metro (stops)

### DISCUSSION THREAD

**Discussion by rafipiccolo** [transitland/transitland#411] [Q&A] [answered]
We use V1 of the API to retrieve the list of metro / tram and train based on a latitude and longitude.

https://transit.land/api/v1/stops?lon=${lng}&lat=${lat}&r=1000&served_by_vehicle_types[]=metro&served_by_vehicle_types[]=tram&served_by_vehicle_types[]=rail&apikey=

In V2, we can't find a way to filter the call by "route_type" : https://www.transit.land/documentation/rest-api/stops

Can you help us ?

Another question :

On V1 for each call, we had this data in every "stop" : 

"routes_serving_stop": [
                {
                    "operator_name": "Tramway",
                    "operator_onestop_id": "o-u09w-tramway",
                    "route_name": "T1",
                    "route_onestop_id": "r-u09w-t1"
                }
            ],

"operators_serving_stop": [
                {
                    "operator_name": "Tramway",
                    "operator_onestop_id": "o-u09w-tramway"
                }
            ],

"served_by_vehicle_types": [
                "tram"
            ],

We can't find this information in the V2 answer.

tldr: How can we list metro / tram and train stations with V2.

**Comment by drewda** [comment, marked as answer]
Thanks for working to migrating your code from v1 to v2 REST.

You're right that our team had forgotten to allow filtering of stops by the route vehicle types that visit those stops. We've just rolled out an addition. See the new `served_by_route_type` query param on the [stops endpoint](https://www.transit.land/documentation/rest-api/stops).

Regarding your second question, the v2 REST API is much more performant than the v1 Datastore API because it constrains some of the tables that are joined and queried behind the scenes. We know this means that some applications may need to make multiple requests to assemble some of the data they could previously receive in one request from v1. But the benefits in performance for clients, servers, and database should be useful for all.

**Comment by rafipiccolo** [comment]
thx for the filter we saw this in the doc to define the route types : (correct ?)
```
0 - Tram, Streetcar, Light rail. Any light rail or street level system within a metropolitan area.
1 - Subway, Metro. Any underground rail system within a metropolitan area.
2 - Rail. Used for intercity or long-distance travel.
```
so we tried this :
```
curl 'https://transit.land/api/v2/rest/stops?lon=2.3301756&lat=48.9331046&radius=1000&apikey=xxxxxxxxxxxxxx&limit=100&served_by_route_type[]=0&served_by_route_type[]=1&served_by_route_type[]=2' | jq
```
and it doenst seem like it filters anything.


also it returned a lot of results. here is a sample. We still have no clue how to get the route type now from this list.
we typically want to know that there is a tram named "T1", that serves the station "La noue". should we make another call ? to what ?
```
{
  "stops": [
    {
      "feed_version": {
        "feed": {
          "id": 762,
          "onestop_id": "f-u09-paris~metro~bus~tram"
        },
        "fetched_at": "2023-09-21T03:12:20.26872Z",
        "id": 357155,
        "sha1": "4e5640d6a521f05dc13c9ca189e9a01cdc9fcd1b"
      },
      "geometry": {
        "coordinates": [
          2.327495770159331,
          48.93479678750794
        ],
        "type": "Point"
      },
      "id": 646522592,
      "level": null,
      "location_type": 1,
      "onestop_id": "s-u09wkuqcfe-lanoue",
      "parent": null,
      "place": {
        "adm0_iso": "FR",
        "adm0_name": "France",
        "adm1_iso": "FR-92",
        "adm1_name": "Hauts-de-Seine"
      },
      "platform_code": null,
      "stop_code": "",
      "stop_desc": "",
      "stop_id": "IDFM:72313",
      "stop_name": "La Noue",
      "stop_timezone": "",
      "stop_url": "",
      "tts_stop_name": null,
      "wheelchair_boarding": 0,
      "zone_id": ""
    }
...
```

**Reply by drewda** [reply]
The new `served_by_route_type` query parameter on the v2 REST stops endpoint takes only one value at a time. It cannot take an array of values. Please try revising your query to issue one request per route type.

**Reply by rafipiccolo** [reply]
ok i can do that.

please note that v1 was nearly instant and had complete answers.
v2 is 4 minute per call and i have to do 3 of them followed by x subcalls to get metro/tram name for each stop :)
i hope it's only a ressource provisioning issue.

nevertheless, how can i get the metro name please ?
i think i tried every possible id

**Reply by rafipiccolo** [reply]
@drewda sorry to insist but how to get the metro name from stop_id ? or anything else ?

**Reply by farfromrefug** [reply]
I face the same slowness issue with v2. It is slower than v1 and having to make multiple requests instead of having a parameter supporting multiple values makes it worse(and I dont see how it could be lighter server side(. The consequence is a bad user experience which is the worst case

**Reply by Alexiscss** [reply]
Hello @drewda as mentioned by @rafipiccolo could you tell us how can i get the metro name ?
Thank you in advance for your time

**Reply by rafipiccolo** [reply]
OK, so we found a way now.

we need to call routes like so
curl 'https://transit.land/api/v2/rest/routes.json?apikey=xxx&lon=2.2736366921941396&lat=48.824574107789964&radius=100' | jq

and then for every result get the corresponding stops : by filtering with served_by_onestop_ids
curl 'https://transit.land/api/v2/rest/stops.json?apikey=xxxxxxxx&lon=2.2736366921941396&lat=48.824574107789964&radius=1000&served_by_onestop_ids=r-u09t-12' | jq

it's incredibly slow compared to v1

**Reply by NichitaCon** [reply]
is there still no way to get vehicle types in one api call?

**Comment by Alexiscss** [comment]
Hi, 

Is there any way to make it work like before ? 

Sincerely, 


A.

Centralisez vos outils & optimisez votre efficacité 🚀
Alexis COISSAC CEO & Co-founder
***@***.*** ***@***.***>
+33 7 64 07 01 08 <tel:+33 7 64 07 01 08>
👉 Réserver une démo <https://calendly.com/alexiscss-86/45min>
